Investigate store-gateway panic

On weekly-f50 I have observed this panic:

```
2024-01-22 05:28:36.376 (UTC)

	github.com/grafana/pyroscope/pkg/storegateway/bucket_stores.go:248 +0x650
created by github.com/grafana/pyroscope/pkg/storegateway.(*BucketStores).syncUsersBlocks in goroutine 48
	github.com/grafana/pyroscope/pkg/storegateway/bucket_stores.go:252 +0x134
github.com/grafana/pyroscope/pkg/storegateway.(*BucketStores).syncUsersBlocks.func2()
	github.com/grafana/pyroscope/pkg/storegateway/bucket_stores.go:166 +0x2c
github.com/grafana/pyroscope/pkg/storegateway.(*StoreGateway).syncStores.(*BucketStores).SyncBlocks.func1({0x2c438b0?, 0xc0000f2af0?}, 0x4c862512022e4?)
	github.com/grafana/pyroscope/pkg/storegateway/bucket.go:108 +0x4e
github.com/grafana/pyroscope/pkg/storegateway.(*BucketStore).SyncBlocks(0xc000c9ef30, {0x2c438b0?, 0xc0000f2af0})
	github.com/grafana/pyroscope/pkg/storegateway/bucket_index_metadata_fetcher.go:108 +0x613
github.com/grafana/pyroscope/pkg/storegateway.(*BucketIndexMetadataFetcher).Fetch(0xc00077a1c0, {0x2c438b0, 0xc0000f2af0})
goroutine 18519010 [running]:
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x1af4a93]
panic: runtime error: invalid memory address or nil pointer dereference
```

--- a/pkg/storegateway/bucket_index_metadata_fetcher.go
+++ b/pkg/storegateway/bucket_index_metadata_fetcher.go
@@ -103,6 +103,12 @@ func (f *BucketIndexMetadataFetcher) Fetch(ctx context.Context) (metas map[ulid.
 
 		return f.fallbackFetch(ctx)
 	}
+	if err != nil {
+		f.metrics.Synced.WithLabelValues(block.FailedMeta).Set(1)
+		f.metrics.Submit()
+
+		return nil, nil, errors.Wrapf(err, "read bucket index")
+	}
 
 	// check if index is older than 1 hour, fallback to metafetcher
 	if time.Unix(idx.UpdatedAt, 0).Before(start.Add(-1 * time.Hour)) {
@@ -122,13 +128,6 @@ func (f *BucketIndexMetadataFetcher) Fetch(ctx context.Context) (metas map[ulid.
 		}
 	}()
 	f.metrics.Syncs.Inc()
-
-	if err != nil {
-		f.metrics.Synced.WithLabelValues(block.FailedMeta).Set(1)
-		f.metrics.Submit()
-
-		return nil, nil, errors.Wrapf(err, "read bucket index")
-	}
 
 	// Build block metas out of the index.
 	metas = make(map[ulid.ULID]*block.Meta, len(idx.Blocks))

